### PR TITLE
#2824: Automatically add to DocBlock comments the thrown `\Throwables`.

### PR DESCRIPTION
**Pull request by Aerendir**
#2824: Automatically add to DocBlock comments the thrown `\Throwables`.

Was PR https://github.com/rectorphp/rector/pull/2828

- [x] Support methods
- [x] Support functions

# Todo in next release
- [ ] Support `\Throwables` assigned to variables (`$value = new \Exception()`)[https://github.com/rectorphp/rector/pull/2833#discussion_r377556485]
- [ ] Support `\Throwables` generated by methods or functions (`throw $this->createException()` and `throw createException()`)
- [ ] Aliased imports (two exceptions with the same name, but different namespaces and one of them is aliased)

### CODE REVIEW

**Comment by Aerendir** [comment]
@TomasVotruba , I did a mess trying to sync the branch with master and I recreated it from scratch.

It seems that it doesn't work anymore: the tests that once passed, now don't pass anymore.

It is seems that the modifications are not get by Rector.

The `$node` I return seems to be modified with the update DocComment, but the test fails anyway.

Any ideas about why?

**Comment by TomasVotruba** [comment]
Can you allow me to push to the branch?

**Comment by TomasVotruba** [comment]
I don't know how to push to your external branch.
I'll make you a collab, so we can cooperate more easily

**Comment by TomasVotruba** [comment]
I sent you an invite

**Comment by Aerendir** [comment]
@TomasVotruba in the mean time I understood how to use `AttributeKey::PHP_DOC_INFO` so I'm now able to update the doc comment: the test now pass! 🎉

I don't know exactly what happened, but for sure the failing of the test was related to the fact I "hardly" updated the docblock comment without using the `AttributeKey::PHP_DOC_INFO`.

Now I'm going to check again the other tests that once passed...

**Comment by TomasVotruba** [comment]
Great, I was making working demo for php doc info. It's only 5 lines instead of 40 :)

I'm glad you made it work! Could you push it?

**Comment by Aerendir** [comment]
@TomasVotruba just pushed!

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php [hunk 1]
```diff
@@ -98,11 +120,12 @@ private function isDocCommentAlreadyPresent(Doc $docComment, string $pattern):bo
...
      * @param string $FQNOrThrowableName
      *
-     * @return string
+     * @return AttributeAwarePhpDocTagNode
      */
-    private function buildThrowsDocComment(string $FQNOrThrowableName):string
+    private function buildThrowsDocComment(string $FQNOrThrowableName):AttributeAwarePhpDocTagNode
     {
-        return sprintf('@throws %s', $FQNOrThrowableName);
+        $value = new AttributeAwareGenericTagValueNode($FQNOrThrowableName);
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]: This should be probably `AttributeAwareIdentifierTypeNode`

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    /**
+     * @param Throw_ $node
+     *
+     * @return bool
+     */
+    private function isThrowableAnnotated(Throw_ $node) : bool
+    {
+        $method = $node->getAttribute(AttributeKey::METHOD_NODE);
+        $throwableParts = $node->expr->class->parts;
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]:
```suggestion
        $className = $this->getName($node->expr->class);
```

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * @param Throw_ $node
+     *
+     * @return void
+     */
+    private function annotateMethod(Throw_ $node):void
+    {
+        /** @var ClassMethod $method */
+        $method = $node->getAttribute(AttributeKey::METHOD_NODE);
+        $throwableParts = $node->expr->class->parts;
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]:
```suggestion
        $className = $this->getName($node->expr->class);
```

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php (hunk 1 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]: I don't know the difference, honestly 😅
I'll change it as you suggest.

**Comment by Aerendir** [comment]
Just run `composer complete-check`: there are some tests that fails that are not related to my PR.

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            return  true;
+        }
+
+        foreach ($comments as $comment) {
+            if ($this->isDocCommentAlreadyPresent($comment, $pattern)) {
+                return true;
+            }
+        }
+
+        return false;
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]: You can use `PhpDocInfo` here too:

```php
/** @var PhpDocInfo $phpDocInfo */
        $phpDocInfo = $throw->getAttribute(AttributeKey::PHP_DOC_INFO);
        $throwTags = $phpDocInfo->getTagsByName('throws');

        /** @var ThrowsTagValueNode $throwTag */
        foreach ($throwTags as $throwTag) {
            // compare...
            $throwTag->type 
        }
        
```

Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]: Ok, I'm going to work just on this to fix the tests.

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php (hunk 1 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowables.php]: `AttributeAwareIdentifierTypeNode` seems to not be compatible with `AttributeAwarePhpDocTagNode`.

Can you tell me what should have I to use?

**Comment by TomasVotruba** [comment]
> Just run composer complete-check: there are some tests that fails that are not related to my PR.

Have you run `composer update`? You might have old deps

**Comment by Aerendir** [comment]
@TomasVotruba , I think the PR is complete.

I'm fixing the checks in CI.

Can you, please, check the code is ok?

**Comment by Aerendir** [comment]
@TomasVotruba , I've fixed as much errors as I can, but there are some for which I need your guidance as they are strictly related to Rector and I think there is already a "standard" approach to fix them.

I'm waiting for your feedback.

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+        return $node;
+    }
+
+    private function isThrowableAnnotated(Throw_ $node): bool
+    {
+        $method = $node->getAttribute(AttributeKey::METHOD_NODE);
+
+        if ($method === null) {
+            throw new RuntimeException('This should not happen and is probably a bug. Please report it.');
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: There is `ShouldNotHappenException` for this

File: packages/coding-style/tests/Rector/ClassMethod/AnnotateThrowablesRector/Fixture/custom_exception_already_annotated.php.inc [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+namespace Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Fixture;
+
+class CustomExceptionAlreadyAnnotated
+{
+    /**
+     * @throws \Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheException
+     */
+    public function throwException()
+    {
+        throw new \Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheException();
```
Comment (TomasVotruba) [inline packages/coding-style/tests/Rector/ClassMethod/AnnotateThrowablesRector/Fixture/custom_exception_already_annotated.php.inc]: Add also new test case fixture with:

```php
throw $value
```

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                return true;
+            }
+
+            if (
+                ! Strings::contains($throwClassName, '\\') &&
+                Strings::contains($FQN, $throwClassName) &&
+                $this->isThrowableImported($node)
+            ) {
+                return true;
+            }
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: This is really not needed, as the `$this->getName()` method always returns FQN name

Stripping `\\` of docblock should help:

```php
$throwClassName = trim($throwClassName, '\\');
if ($throwClassName === $FQN) {
    return true;
}
```

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            /** @var UseUse $useStmt */
+            foreach ($useNode->uses as $useStmt) {
+                if ($this->getName($useStmt) === $throwClassName) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: Can be removed

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+    private function annotateMethod(Throw_ $node): void
+    {
+        /** @var ClassMethod $method */
+        $method = $node->getAttribute(AttributeKey::METHOD_NODE);
+        $FQN = $this->buildFQN($node);
```
Comment (TomasVotruba) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: This `$FQN` was already resolved before. It shold be in argument of this method.

Also FQN is too vague name. It could be any class. But actually this is `$exceptionClass` or `$throwableClass`, maybe use that instead

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 5 above)
Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: Also, throw can be used in a function as well. There is no `METHOD_NODE`.
It would make more sense to `return null;` instead

**Comment by Aerendir** [comment]
@TomasVotruba , I read all your comments: I will work on them ASAP.

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 5 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: I didn't took functions into account at all.

Do you think we should first release and then add support for functions or should have I to work first on implementing functions support?

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 7 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: Using

```php
        /** @var AttributeAwarePhpDocTagNode $throwTag */
        foreach ($throwTags as $throwTag) {
            // $thrownClassName = $throwTag->value->type->name;
            $thrownClassName = $this->getName($throwTag);
```

Causes an error:

> TypeError: Argument 1 passed to Rector\Core\Rector\AbstractRector::getName() must implement interface PhpParser\Node, instance of Rector\AttributeAwarePhpDoc\Ast\PhpDoc\AttributeAwarePhpDocTagNode given

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 8 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: Depends on this: https://github.com/rectorphp/rector/pull/2833#discussion_r377557195

**Comment by Aerendir** [comment]
@TomasVotruba , regarding variables and factory method the things are really complex.

Imagine this code:

```php
<?php

function createHttpException(int $statusCode) {
    switch ($statusCode) {
        case 404:
            return new NotFoundException();
        case 500:
            return new InternalServerErrorException();
        default:
            return new \ConnectionException();
    }
}

/**
 * @param string $url
 * 
 * @throws ... What?
 */
function connect(string $url)
{
    // do connection things
    $response = ...;
    
    // Imagine is 404
    $status = $response->getStatus();
    
    if (200 !== $status) {
        throw createHttpException(404);
    }
}
```

What is the thrown exception? All? And for more complex logics? And how can we understand the code to annotate the DocBlock?

I've done an example with factory methods, but the same applies also to `\Throwables` assigned to variables.

WDYT?

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 8 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: Read this: https://github.com/rectorphp/rector/pull/2833#issuecomment-584948298

File: packages/coding-style/tests/Rector/ClassMethod/AnnotateThrowablesRector/Fixture/custom_exception_already_annotated.php.inc (hunk 6 above)
Comment (Aerendir) [reply packages/coding-style/tests/Rector/ClassMethod/AnnotateThrowablesRector/Fixture/custom_exception_already_annotated.php.inc]: Read this: https://github.com/rectorphp/rector/pull/2833#issuecomment-584948298

**Comment by TomasVotruba** [comment]
> What is the thrown exception? All? And for more complex logics? And how can we understand the code to annotate the DocBlock?

I don't use it. What is common approach?

**Comment by TomasVotruba** [comment]
Maybe we could solve complex annotations like this in next PR, so there is smaller scope to focus on and finished work

**Comment by Aerendir** [comment]
> I don't use it. What is common approach?

Ok, let me think a bit about that and so some tests with Psalm and prepare some failing tests...

I'll also test the current version of the Rector on my app to see what happens.

Then we will make a decision.

**Comment by Aerendir** [comment]
It seems that Psalm doesn't issue anything in case of use `throw createHttpException(404)`:

https://psalm.dev/r/44e615c135

**Comment by Aerendir** [comment]
Psalm test: https://psalm.dev/r/5c998a075e

**Comment by TomasVotruba** [comment]
We run on PHPStan, so Psalm test is useless here

**Comment by Aerendir** [comment]
It is a memo to not be forced to rewrite again the code each time.

PHPStan doesn't alert about missing `@throws` tags, so Psalm is useful to me to understand its behaviour beacuse this helps me in finding the behavior to give to the rector.

Having the snippet linked here makes me able to edit it instead of rewriting it each time and fastly clarify what happens in different situations instead of having a project on my computer just to test a bunch of lines of code.

I hope you understand my point of view.

**Comment by Aerendir** [comment]
This is the same code run on PHPStan: https://phpstan.org/r/c6491906-2837-44fa-b0c7-8400338f977a

No alert about missed `@throws` tags :(

**Comment by TomasVotruba** [comment]
I understand the idea. Just that you'll have to figure out how to make that work with Rector and PHPStan anyway, so Psalm might lead you astrey.

**Comment by Aerendir** [comment]
@TomasVotruba , I'm playing with PhpParser since two or three hours, but I cannot find a way to identify a call to a factory method.

**The main point is that to go forward with the implementation I need to identify and distinguish these two cases:**

```php
throw new \Exception();
throw ExceptionsFactoryNotAnnotated::createException(1)
```

This is the failing test case:

```php
<?php

namespace Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Fixture;

use Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheException;
use Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheExceptionTheSecond;
use Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheExceptionTheThird;

class ExceptionsFactoryNotAnnotated
{
    public static function createException(int $code)
    {
        switch ($code) {
            case 1:
                return new TheException();
            case 2:
                return new TheExceptionTheSecond();
            case 3:
                return new TheExceptionTheThird();
            default:
                return new \RuntimeException();
        }
    }
}

function throwWithFactoryMethodNotAnnotated()
{
    throw ExceptionsFactoryNotAnnotated::createException(1);
}

?>
-----
<?php

namespace Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Fixture;

use Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheException;
use Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheExceptionTheSecond;
use Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Source\TheExceptionTheThird;

class ExceptionsFactoryNotAnnotated
{
    public static function createException(int $code)
    {
        switch ($code) {
            case 1:
                return new TheException();
            case 2:
                return new TheExceptionTheSecond();
            case 3:
                return new TheExceptionTheThird();
            default:
                return new \RuntimeException();
        }
    }
}

function throwWithFactoryMethodNotAnnotated()
{
    throw ExceptionsFactoryNotAnnotated::createException(1);
}

?>

```

This is what the current implementation outputs:

```diff
--- Expected
+++ Actual
@@ @@
     }
 }
 
+/**
+ * @throws \Rector\CodingStyle\Tests\Rector\ClassMethod\AnnotateThrowables\Fixture\ExceptionsFactoryNotAnnotated
+ */
 function throwWithFactoryMethodNotAnnotated()
 {
     throw ExceptionsFactoryNotAnnotated::createException(1);

```

**In this case, as there is no return type hinting nor any `@return` Docblock tag, the `$node` should not be modified** (at least Psalm warns about the missing return type hint or about the missing `@return` Docblock tag... but this is another story).

**The main point is that to go forward with the implementation I need to identify and distinguish these two cases:**

```php
throw new \Exception();
throw ExceptionsFactoryNotAnnotated::createException(1)
```

Any ideas about how to that?

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+            default:
+                dumpf(null, $node);
+                throw new ShouldNotHappenException(sprintf('The \Throwable %s is not supported. Please, open an issue.', get_class($node->expr)));
+        }
+    }
+
+    private function identifyThrownThrowablesInStaticCall(Throw_ $node):array
+    {
+        ddf(null, 'Nodo' , $node);
+    }
```
Comment (Aerendir) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: @TomasVotruba , Any ideas about the path to follow to access the dockblock and return type hint of the static method?

Failing test case: https://github.com/rectorphp/rector/pull/2833/files#diff-d1249893d1b00c7b6078b3d7924b48d4

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 7 above)
Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: That's doc block tag, it's something else than php-parser node

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 10 above)
Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: I don't understand what is needed.

Could you provide minimal PHP code snippet that explain what should return when?

File: composer.json [hunk 11]
```diff
@@ -39,6 +39,7 @@
         "jetbrains/phpstorm-stubs": "^2019.2",
         "ocramius/package-versions": "^1.4|^1.5",
         "phpunit/phpunit": "^8.4|^9.0",
+        "serendipity_hq/vardumper-cli-to-html": "^1.0",
```
Comment (TomasVotruba) [inline composer.json]: Try `dump()`, it's inclued with Tracy

Comment (Aerendir) [reply composer.json]: Doesn't `dump()` conflicts with the one from Symfony's VarDumper component?

I wrote yesterday this small function to save the dumped result to an HTML file that makes easier the navigation across nodes.

Comment (Aerendir) [reply composer.json]: Anyway I will remove it once I will have solved the problem of finding a path to access the static method's docblocks and return type hints.

Comment (TomasVotruba) [reply composer.json]: > Doesn't dump() conflicts with the one from Symfony's VarDumper component?

It's possible. In that case we should have own `dump()` function that is loaded first and delegates to Tracy one.

Symfony is not good in CLI

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 10 above)
Comment (Aerendir) [inline packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: The full code is this.

If I find a `New_` node, then it is a throw like `throw new \Exception()`: non problem in this case.

But if the exception is thrown as `throw ExceptionFactory::createException()`, then I need to access the `createException()` method and read its docblock to see if the return type is annotated and, if not, read its return type hint.
The results of the Docblock `@return` annotation or, alternatively, of the return type hint have to then be used to annotate the method/function to which the analyzing `Throw_` node belongs to.

In case of static methods, I get a `StaticCall` node: from this I cannot find a path to access the node that pertains to `ExceptionFactory::createException()`: I can read the class name ´ExceptionFactory` and the method name `createException`, but cannot access their nodes.

Without being able to access their nodes I cannot know what is returned by `ExceptionFactory::createException()` and so I cannot annotate the `Throw_` node.

Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: In that case, I'd try `$this->getType($node)`

Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: @TomasVotruba , if you download the PR (did you see this https://github.blog/2020-02-12-supercharge-your-command-line-experience-github-cli-is-now-in-beta/ ?), you can better understand which is the problem.

If you run `vendor/bin/phpunit --filter AnnotateThrowables`, the `dumpf()` call will create an HTML file useful to navigate the full `Throw_ $node`: you will see it seems there is no path toward the `ExceptionFactory::createException()` node... or at least I didn't found it 😅

Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: For me, the PR is getting to huge to work effectively. We usually have 200-300 lines tests included.

I'd like to merge basic feature now to `master`, then resolve the edge cases that don't pass yet.

Could you move those non passing fixtures to `/FixtureTodo` diretory?

Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: I'll check the Github stuff, looks neat, thanks :+1:

File: composer.json (hunk 11 above)
Comment (Aerendir) [reply composer.json]: @TomasVotruba 
> Symfony is not good in CLI

Exactly: it is impossible to read in CLI.

So I created this small function that saves the dump to an HTML file and so you have the dump from the CLI, but as if it was triggered in an HTML page: this way you can open and close nodes, highlight same objects in different path, search in the dump, etc...

File: packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php (hunk 10 above)
Comment (Aerendir) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: > For me, the PR is getting to huge to work effectively. We usually have 200-300 lines tests included.
> 
> I'd like to merge basic feature now to `master`, then resolve the edge cases that don't pass yet.
> 
> Could you move those non passing fixtures to `/FixtureTodo` diretory?

Ok, let me play for a bit more time: tomorrow I will send a working PR with the working rector.

Comment (TomasVotruba) [reply packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php]: I'll be ready to merge :)

**Comment by Aerendir** [comment]
@TomasVotruba , I reverted the branch to a stable state: check if it is ready to be merged.

Then I will open other PRs to implement the other features...

**Comment by Aerendir** [comment]
PS
I've also added the folder `Throw_` as actually the rector doesn't operate on methods anymore, but also on functions.

**Comment by TomasVotruba** [comment]
CI needs to pass though

**Comment by Aerendir** [comment]
> CI needs to pass though

It doesn't trigger as I updated the `docs/AllRectorsOverview.md` and it now conflicts with the one in `master` (another rector added?)...

**Comment by TomasVotruba** [comment]
Just change some space/file and commit

**Comment by TomasVotruba** [comment]
Rebase on `master` is needed too. That's why it's better to create small PR and merge it in 1-2 days then re-iterate, that do this tedius tyding up work.

**Comment by Aerendir** [comment]
@TomasVotruba , I've fixed what I can, but there is the conflict on `docs/AllRectorsOverview.md` that I don't know how to solve another issue with PHPStan that complains about the line 110 (`return $alreadyAnnotatedThrowTag->value->type->name;`) saying that:

```console
vendor/bin/phpstan analyse --ansi --error-format symplify
Note: Using configuration file /Users/Aerendir/Documents/JooServer/_ProjectsIContributeTo/rector/phpstan.neon.
 1602/1602 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------------------------------------------------------------------------
 packages/coding-style/src/Rector/Throw_/AnnotateThrowablesRector.php:110
 ------------------------------------------------------------------------
 - '#Access to an undefined property PHPStan\\PhpDocParser\\Ast\\PhpDoc\\PhpDocTagValueNode\:\:\$type#'
 ------------------------------------------------------------------------
```

There are also other similar issues...

**Comment by TomasVotruba** [comment]
Just rebuild docs with `composer docs`

**Comment by TomasVotruba** [comment]
The phpstan issue can be solved with `instanceof`

If that's the only blocker, I can merge it and resolve

**Comment by Aerendir** [comment]
> Just rebuild docs with `composer docs`

It didn't worked: I already rebuilt docs, but there is still a conflict...

> The phpstan issue can be solved with `instanceof`

It should...

> If that's the only blocker, I can merge it and resolve

For sure you have more familiarity with the repository 😅

**Comment by Aerendir** [comment]
As told some days ago, on my local fork there are also some failing tests not related to this PR and I don't know how to solve them but, mostly, if I have to...

**Comment by TomasVotruba** [comment]
Is it possible you didn't sync your fork first? 

I'll look on it
